### ISSUE TITLE
fuse_request_end: kernel tried to execute NX-protected page

### ISSUE THREAD

**Issue by krumelmonster** [rfjakob/gocryptfs#458]
For some reason I became unable to change any file attributes on my gocryptfs-ecrypted home directory. I can list and read/write though. The system is up and running and I'll only suspend for the night so we can debug. Also: I'm really worried that I might not be able to mount this fs ever again.

See for example strace touch test
```
execve("/usr/bin/touch", ["touch", "test"], 0x7ffd93762f28 /* 54 vars */) = 0
brk(NULL)                               = 0x55c9c9d69000
arch_prctl(0x3001 /* ARCH_??? */, 0x7ffd2f1c7d40) = -1 EINVAL (Invalid argument)
access("/etc/ld.so.preload", R_OK)      = -1 ENOENT (No such file or directory)
openat(AT_FDCWD, "/etc/ld.so.cache", O_RDONLY|O_CLOEXEC) = 3
fstat(3, {st_mode=S_IFREG|0644, st_size=380366, ...}) = 0
mmap(NULL, 380366, PROT_READ, MAP_PRIVATE, 3, 0) = 0x7f900f1d9000
close(3)                                = 0
openat(AT_FDCWD, "/usr/lib/libc.so.6", O_RDONLY|O_CLOEXEC) = 3
read(3, "\177ELF\2\1\1\3\0\0\0\0\0\0\0\0\3\0>\0\1\0\0\0`q\2\0\0\0\0\0"..., 832) = 832
pread64(3, "\6\0\0\0\4\0\0\0@\0\0\0\0\0\0\0@\0\0\0\0\0\0\0@\0\0\0\0\0\0\0"..., 784, 64) = 784
pread64(3, "\4\0\0\0\20\0\0\0\5\0\0\0GNU\0\2\0\0\300\4\0\0\0\3\0\0\0\0\0\0\0", 32, 848) = 32
pread64(3, "\4\0\0\0\24\0\0\0\3\0\0\0GNU\0\22\344\22\321\223\216\303\377yu\37\16\205\363\33\305"..., 68, 880) = 68
fstat(3, {st_mode=S_IFREG|0755, st_size=2145592, ...}) = 0
mmap(NULL, 8192, PROT_READ|PROT_WRITE, MAP_PRIVATE|MAP_ANONYMOUS, -1, 0) = 0x7f900f1d7000
pread64(3, "\6\0\0\0\4\0\0\0@\0\0\0\0\0\0\0@\0\0\0\0\0\0\0@\0\0\0\0\0\0\0"..., 784, 64) = 784
pread64(3, "\4\0\0\0\20\0\0\0\5\0\0\0GNU\0\2\0\0\300\4\0\0\0\3\0\0\0\0\0\0\0", 32, 848) = 32
pread64(3, "\4\0\0\0\24\0\0\0\3\0\0\0GNU\0\22\344\22\321\223\216\303\377yu\37\16\205\363\33\305"..., 68, 880) = 68
mmap(NULL, 1856312, PROT_READ, MAP_PRIVATE|MAP_DENYWRITE, 3, 0) = 0x7f900f011000
mmap(0x7f900f036000, 1359872, PROT_READ|PROT_EXEC, MAP_PRIVATE|MAP_FIXED|MAP_DENYWRITE, 3, 0x25000) = 0x7f900f036000
mmap(0x7f900f182000, 307200, PROT_READ, MAP_PRIVATE|MAP_FIXED|MAP_DENYWRITE, 3, 0x171000) = 0x7f900f182000
mmap(0x7f900f1cd000, 24576, PROT_READ|PROT_WRITE, MAP_PRIVATE|MAP_FIXED|MAP_DENYWRITE, 3, 0x1bb000) = 0x7f900f1cd000
mmap(0x7f900f1d3000, 13112, PROT_READ|PROT_WRITE, MAP_PRIVATE|MAP_FIXED|MAP_ANONYMOUS, -1, 0) = 0x7f900f1d3000
close(3)                                = 0
mmap(NULL, 8192, PROT_READ|PROT_WRITE, MAP_PRIVATE|MAP_ANONYMOUS, -1, 0) = 0x7f900f00f000
arch_prctl(ARCH_SET_FS, 0x7f900f1d85c0) = 0
mprotect(0x7f900f1cd000, 12288, PROT_READ) = 0
mprotect(0x55c9c98ab000, 4096, PROT_READ) = 0
mprotect(0x7f900f261000, 4096, PROT_READ) = 0
munmap(0x7f900f1d9000, 380366)          = 0
brk(NULL)                               = 0x55c9c9d69000
brk(0x55c9c9d8a000)                     = 0x55c9c9d8a000
openat(AT_FDCWD, "/usr/lib/locale/locale-archive", O_RDONLY|O_CLOEXEC) = 3
fstat(3, {st_mode=S_IFREG|0644, st_size=3385504, ...}) = 0
mmap(NULL, 3385504, PROT_READ, MAP_PRIVATE, 3, 0) = 0x7f900ecd4000
close(3)                                = 0
openat(AT_FDCWD, "test", O_WRONLY|O_CREAT|O_NOCTTY|O_NONBLOCK, 0666) = 3
dup2(3, 0)                              = 0
close(3)                                = 0
utimensat(0, NULL, NULL, 0)             = -1 ENOENT (No such file or directory)
close(0)                                = 0
write(2, "touch: ", 7touch: )                  = 7
write(2, "setting times of 'test'", 23setting times of 'test') = 23
write(2, ": No such file or directory", 27: No such file or directory) = 27
write(2, "\n", 1
)                       = 1
close(1)                                = 0
close(2)                                = 0
exit_group(1)                           = ?
+++ exited with 1 +++
```

mount options are `rw,nosuid,nodev,relatime,default_permissions,allow_other,max_read=131072`
Please excuse my brevity, it is late and I have to take care of running the backups.

**Comment by rfjakob** [comment]
Interesting. Does chmod fail as well?

**Comment by krumelmonster** [comment]
Yes, `fchmodat(AT_FDCWD, "app", 0755) = -1 ENOENT (No such file or directory)`

**Comment by rfjakob** [comment]
Can you run an `strace -f -p` on the gocryptfs PID while you run the chmod?

Also, you could try mounting the encrypted directory a second time on a different mountpoint (this does no harm) and see how things look there.

**Comment by krumelmonster** [comment]
On it now. I just noticed the gocryptfs process is defunct

**Comment by krumelmonster** [comment]
Ok, thanks a lot for your support last night. It's probably first of all good to have this issue as a reference: If the gocryptfs process goes defunct, the fs will keep working except changing file attributes will report `No such file or directory`. I'm not sure if there's something to do about this.
@rfjakob Would you like to investigate further on what caused the process to go defunct? I don't think it's really my fault, I didn't send any signals to it, I just made (heavy?) use of the fs.
coredumpctl doesn't list a dump for gocryptfs. I'd go remount the fs now if you don't mind?

**Comment by rfjakob** [comment]
Before you remount, please capture `ps axjf`. It's not plausible that the fs keeps working when the gocryptfs process shows up as defunct. Maybe one thread died and the others are still working.

**Comment by rfjakob** [comment]
Hmm, ps axjf does not show all threads. Give me a second.

**Comment by rfjakob** [comment]
This seems ok, no forest display, though:

`ps auxwww -T`

**Comment by krumelmonster** [comment]
```
$ ps auxwww -T | grep gocryptfs
jo           705     705  0.0  0.0      0     0 ?        Zsl  Mar03   5:37 [gocryptfs] <defunct>
jo           705     706  0.1  4.7 3532744 565572 ?      Ssl  Mar03   7:39 /usr/bin/gocryptfs -fg -notifypid=698 -nodev -nosuid -quiet -nonempty -allow_other /home/jo.cipher /home/jo
jo           705     707  0.0  4.7 3532744 565572 ?      Ssl  Mar03   0:00 /usr/bin/gocryptfs -fg -notifypid=698 -nodev -nosuid -quiet -nonempty -allow_other /home/jo.cipher /home/jo
jo           705     708  0.0  4.7 3532744 565572 ?      Ssl  Mar03   0:00 /usr/bin/gocryptfs -fg -notifypid=698 -nodev -nosuid -quiet -nonempty -allow_other /home/jo.cipher /home/jo
jo           705     709  0.0  4.7 3532744 565572 ?      Ssl  Mar03   5:21 /usr/bin/gocryptfs -fg -notifypid=698 -nodev -nosuid -quiet -nonempty -allow_other /home/jo.cipher /home/jo
jo           705     710  0.0  4.7 3532744 565572 ?      Ssl  Mar03   0:00 /usr/bin/gocryptfs -fg -notifypid=698 -nodev -nosuid -quiet -nonempty -allow_other /home/jo.cipher /home/jo
jo           705     714  0.1  4.7 3532744 565572 ?      Ssl  Mar03   7:12 /usr/bin/gocryptfs -fg -notifypid=698 -nodev -nosuid -quiet -nonempty -allow_other /home/jo.cipher /home/jo
jo           705     715  0.0  4.7 3532744 565572 ?      Ssl  Mar03   6:13 /usr/bin/gocryptfs -fg -notifypid=698 -nodev -nosuid -quiet -nonempty -allow_other /home/jo.cipher /home/jo
jo           705     717  0.1  4.7 3532744 565572 ?      Ssl  Mar03   9:27 /usr/bin/gocryptfs -fg -notifypid=698 -nodev -nosuid -quiet -nonempty -allow_other /home/jo.cipher /home/jo
jo           705     814  0.0  4.7 3532744 565572 ?      Ssl  Mar03   4:46 /usr/bin/gocryptfs -fg -notifypid=698 -nodev -nosuid -quiet -nonempty -allow_other /home/jo.cipher /home/jo
jo           705     819  0.0  4.7 3532744 565572 ?      Ssl  Mar03   2:05 /usr/bin/gocryptfs -fg -notifypid=698 -nodev -nosuid -quiet -nonempty -allow_other /home/jo.cipher /home/jo
jo           705   26589  0.0  4.7 3532744 565572 ?      Ssl  Mar03   0:00 /usr/bin/gocryptfs -fg -notifypid=698 -nodev -nosuid -quiet -nonempty -allow_other /home/jo.cipher /home/jo
jo           705   26590  0.0  4.7 3532744 565572 ?      Ssl  Mar03   4:11 /usr/bin/gocryptfs -fg -notifypid=698 -nodev -nosuid -quiet -nonempty -allow_other /home/jo.cipher /home/jo
jo           705   26592  0.0  4.7 3532744 565572 ?      Ssl  Mar03   4:13 /usr/bin/gocryptfs -fg -notifypid=698 -nodev -nosuid -quiet -nonempty -allow_other /home/jo.cipher /home/jo
jo           705   29041  0.1  4.7 3532744 565572 ?      Ssl  Mar03   7:04 /usr/bin/gocryptfs -fg -notifypid=698 -nodev -nosuid -quiet -nonempty -allow_other /home/jo.cipher /home/jo
jo           705   29042  0.0  4.7 3532744 565572 ?      Ssl  Mar03   0:00 /usr/bin/gocryptfs -fg -notifypid=698 -nodev -nosuid -quiet -nonempty -allow_other /home/jo.cipher /home/jo
jo           705   29043  0.1  4.7 3532744 565572 ?      Ssl  Mar03   6:45 /usr/bin/gocryptfs -fg -notifypid=698 -nodev -nosuid -quiet -nonempty -allow_other /home/jo.cipher /home/jo
jo           705   29059  0.0  4.7 3532744 565572 ?      Ssl  Mar03   0:53 /usr/bin/gocryptfs -fg -notifypid=698 -nodev -nosuid -quiet -nonempty -allow_other /home/jo.cipher /home/jo
jo           705   29060  0.1  4.7 3532744 565572 ?      Ssl  Mar03   9:00 /usr/bin/gocryptfs -fg -notifypid=698 -nodev -nosuid -quiet -nonempty -allow_other /home/jo.cipher /home/jo
jo           705   29061  0.1  4.7 3532744 565572 ?      Ssl  Mar03   7:05 /usr/bin/gocryptfs -fg -notifypid=698 -nodev -nosuid -quiet -nonempty -allow_other /home/jo.cipher /home/jo
jo           705   29062  0.1  4.7 3532744 565572 ?      Ssl  Mar03   8:16 /usr/bin/gocryptfs -fg -notifypid=698 -nodev -nosuid -quiet -nonempty -allow_other /home/jo.cipher /home/jo
jo           705   29063  0.0  4.7 3532744 565572 ?      Ssl  Mar03   4:35 /usr/bin/gocryptfs -fg -notifypid=698 -nodev -nosuid -quiet -nonempty -allow_other /home/jo.cipher /home/jo
jo           705   29064  0.0  4.7 3532744 565572 ?      Ssl  Mar03   0:00 /usr/bin/gocryptfs -fg -notifypid=698 -nodev -nosuid -quiet -nonempty -allow_other /home/jo.cipher /home/jo
jo           705  579338  0.0  4.7 3532744 565572 ?      Ssl  Mar05   4:57 /usr/bin/gocryptfs -fg -notifypid=698 -nodev -nosuid -quiet -nonempty -allow_other /home/jo.cipher /home/jo
jo           712     712  0.0  0.0   8088     0 ?        S    Mar03   0:00 logger -t gocryptfs-705-logger
jo       1570678 1570678  0.0  0.0   6632  2312 pts/15   S+   13:30   0:00 grep gocryptfs
```

**Comment by rfjakob** [comment]
Wow, the main thread is defunct, but the children kept running.

All warnings, messages, panics, backtraces should go to syslog (that's what the logger process is for). You should see them like that:

    journalctl --since 2020-03-07 | grep gocryptfs

**Comment by krumelmonster** [comment]
```
Mar 05 15:41:29 myhostname gocryptfs[705]: go-fuse: writer: Write/Writev failed, err: 2=no such file or directory. opcode: INTERRUPT
Mar 05 16:37:58 myhostname gocryptfs[705]: go-fuse: Unimplemented opcode OPCODE-46
Mar 05 18:40:04 myhostname kernel: CPU: 2 PID: 818 Comm: gocryptfs Tainted: G        W  OE     5.5.7-arch1-1 #1
Mar 05 20:26:13 myhostname kernel: [    705]  1000   705   675119     9068   483328      725             0 gocryptfs
Mar 05 20:26:44 myhostname kernel: [    705]  1000   705   675119     9338   483328      730             0 gocryptfs
Mar 07 00:17:55 myhostname gocryptfs[705]: OpenDir "gUKgQUBfb71e0dtRIPDsZA": could not read gocryptfs.diriv: permission denied
Mar 07 00:17:55 myhostname gocryptfs[705]: OpenDir "NjuOrajP5CwaN_efzNsqcw": could not read gocryptfs.diriv: permission denied
Mar 07 00:17:55 myhostname gocryptfs[705]: OpenDir "g91OYU4p5TRtSYvWKML-Eg": could not read gocryptfs.diriv: permission denied
Mar 07 00:17:55 myhostname gocryptfs[705]: OpenDir "z5K5dTDA7zEVKEuse4-gLQ": could not read gocryptfs.diriv: permission denied
Mar 07 15:59:04 myhostname gocryptfs[705]: go-fuse: writer: Write/Writev failed, err: 2=no such file or directory. opcode: INTERRUPT
Mar 07 17:00:37 myhostname gocryptfs[705]: go-fuse: writer: Write/Writev failed, err: 2=no such file or directory. opcode: INTERRUPT
Mar 07 17:16:42 myhostname kernel: CPU: 0 PID: 705 Comm: gocryptfs Tainted: G      D W  OE     5.5.7-arch1-1 #1
Mar 07 23:48:21 myhostname kernel: CPU: 2 PID: 579339 Comm: gocryptfs Tainted: G      D W  OE     5.5.7-arch1-1 #1
Mar 08 00:50:02 myhostname dbus-daemon[519]: [system] Activating via systemd: service name='org.freedesktop.hostname1' unit='dbus-org.freedesktop.hostname1.service' requested by ':1.30652' (uid=1000 pid=1461652 comm="/usr/bin/python /usr/bin/meld stracegocryptfs /tmp")
Mar 08 01:33:18 myhostname kernel: CPU: 1 PID: 29058 Comm: gocryptfs Tainted: G      D W  OE     5.5.7-arch1-1 #1
```

**Comment by krumelmonster** [comment]
```
[147470.306227] kernel tried to execute NX-protected page - exploit attempt? (uid: 1000)
[147470.306232] BUG: unable to handle page fault for address: ffff89e7b8064600
[147470.306234] #PF: supervisor instruction fetch in kernel mode
[147470.306236] #PF: error_code(0x0011) - permissions violation
[147470.306238] PGD 1fde01067 P4D 1fde01067 PUD 1fde06067 PMD 2f80db063 PTE 80000002f8064063
[147470.306242] Oops: 0011 [#2] PREEMPT SMP PTI
[147470.306245] CPU: 0 PID: 705 Comm: gocryptfs Tainted: G      D W  OE     5.5.7-arch1-1 #1

**REMOVED**

[147470.306270] Call Trace:
[147470.306280]  ? fuse_request_end+0xae/0x1e0 [fuse]
[147470.306286]  ? fuse_dev_do_write+0x25d/0xee0 [fuse]
[147470.306292]  ? generic_perform_write+0x149/0x1e0
[147470.306295]  ? wait_woken+0x70/0x70
[147470.306301]  ? fuse_dev_write+0x66/0xa0 [fuse]
[147470.306305]  ? new_sync_write+0x16a/0x210
[147470.306309]  ? vfs_write+0xb6/0x1a0
[147470.306311]  ? ksys_write+0x67/0xe0
[147470.306315]  ? do_syscall_64+0x4e/0x150
[147470.306319]  ? entry_SYSCALL_64_after_hwframe+0x44/0xa9
[147470.306323] Modules linked in: **REMOVED**
```

**Comment by rfjakob** [comment]
Removing the register dump is not a bad idea, not sure what data can be in there.

This looks like a kernel bug or memory corruption (hardware problem). What kernel version do you run?

**Comment by krumelmonster** [comment]
```
Linux karindus 5.5.7-arch1-1 #1 SMP PREEMPT Sat, 29 Feb 2020 19:06:02 +0000 x86_64 GNU/Linux
```
I take back edit 3. I think the timestamps in dmesg -T are wrong (I've heard this is to be expected, esp. when suspening).

**Comment by krumelmonster** [comment]
```
Linux myhostname 5.5.7-arch1-1 #1 SMP PREEMPT Sat, 29 Feb 2020 19:06:02 +0000 x86_64 GNU/Linux
```
I take back edit 3. I think the timestamps in dmesg -T are wrong (I've heard this is to be expected, esp. when suspening).

**Comment by krumelmonster** [comment]
Anything else we should check before I reboot?

**Comment by rfjakob** [comment]
No, reboot at will

**Comment by krumelmonster** [comment]
Am I correct in assuming that this cannot (solely) be a gocryptfs issue because userspace should never be able to cause access violation on the kernel end?
Any recommendations on where to take this issue to get to the bottom of it?

**Comment by rfjakob** [comment]
Yes, correct. I dug a little, fortunately, you are not the first one hitting this bug:

2019-12-08: https://github.com/veracrypt/VeraCrypt/issues/567
2019-12-24: https://github.com/trapexit/mergerfs/issues/686
2019-12-24: https://lkml.org/lkml/2019/12/24/375
2020-01-04: https://www.spinics.net/lists/kernel/msg3363548.html

Looking at https://github.com/trapexit/mergerfs/issues/686#issuecomment-590487731 , there seems to already be a fix queued.

**Comment by rfjakob** [comment]
If I understand correctly, Arch now has kernel 5.5.11
( https://www.archlinux.org/packages/core/x86_64/linux/ )

The fix was merged in Linux v5.5.10 (see https://github.com/trapexit/mergerfs/issues/686#issuecomment-604101915).

So if you run the new kernel, you are safe.
